source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRenderer.mm

```diff
@@ -1,130 +1,35 @@
-{
-    std::vector<std::shared_ptr<BaseCardElement>> body = adaptiveCard->GetBody();
-
-    ACRColumnView *verticalView = nil;
-
-    if(!body.empty())
-    {
-         verticalView = [[ACRColumnView alloc] initWithFrame:CGRectMake(0, 0, guideFrame.size.width, guideFrame.size.height)];
-
-        [ACRRenderer render:verticalView inputs:inputs withCardElems:body andHostConfig:config];
-
-        std::vector<std::shared_ptr<BaseActionElement>> actions = adaptiveCard->GetActions();
-        UIView<ACRIContentHoldingView> *actionChildView = [ACRRenderer renderButton:vc inputs:inputs superview:verticalView actionElems:actions hostConfig:config];
-        [verticalView addArrangedSubview:actionChildView];
-    }
-    return verticalView;
-}
-
-+ (UIView *)renderWithJson:(NSString *)str
-                    inputs:(NSArray *)inputs
-            viewController:(UIViewController *)vc
-                guideFrame:(CGRect)guideFrame
-                hostconfig:(std::shared_ptr<HostConfig> const &)config
+- (instancetype)init
 {
-    std::shared_ptr<AdaptiveCard> adaptiveCard = AdaptiveCard::DeserializeFromString(std::string([str UTF8String]));
-    return [ACRRenderer renderWithAdaptiveCards:adaptiveCard
-                                         inputs:(NSMutableArray *)inputs
-                                 viewController:vc
-                                     guideFrame:guideFrame
-                                     hostconfig:config];
+    self = [super init];
+    return self;
 }
 
-+ (UIView<ACRIContentHoldingView> *)renderButton:(UIViewController *)vc
-                                          inputs:(NSMutableArray *)inputs
-                                       superview:(UIView<ACRIContentHoldingView> *)superview
-                                     actionElems:(std::vector<std::shared_ptr<BaseActionElement>> const &)elems
-                                      hostConfig:(std::shared_ptr<HostConfig> const &)config
++ (ACRRenderResult *)render:(ACOAdaptiveCards *)card config:(ACOHostConfig *)config frame:(CGRect)frame
 {
-    ACRRegistration *reg = [ACRRegistration getInstance];
-    UIView<ACRIContentHoldingView> *childview = nil;
-    UILayoutConstraintAxis axis = UILayoutConstraintAxisVertical;
-    if(ActionsOrientation::Horizontal == config->actions.actionsOrientation)
-    {
-        childview = [[ACRColumnSetView alloc] initWithFrame:CGRectMake(0, 0, superview.frame.size.width, superview.frame.size.height)];
-        axis = UILayoutConstraintAxisHorizontal;
-    }
-    else
-    {
-        childview = [[ACRColumnView alloc] initWithFrame:CGRectMake(0, 0, superview.frame.size.width, superview.frame.size.height)];
-    }
+    ACRRenderResult *result = [[ACRRenderResult alloc] init];
 
-    for(const auto &elem:elems)
-    {
-        ACRBaseActionElementRenderer *actionRenderer =
-        [reg getActionRenderer:[NSNumber numberWithInt:(int)elem->GetElementType()]];
+    if([card isKindOfClass:[ACOAdaptiveCardsInternal class]] == NO)
```

Comment:
>[card isKindOfClass:[ACOAdaptiveCardsInternal class]] [](start = 7, length = 53)

Is there any scenario where there's an ACOAdaptiveCards that _isn't_ an ACOAdaptiveCardsInternal?

If there's only supposed to be one way of instantiating this class, via the fromJson method, then let's not provide an init method at all. I don't think we need the distinction between ACOAdaptiveCards and ACOAdaptiveCardsInternal either.

Reply:
Thanks

---
In reply to: [140325996](https://github.com/Microsoft/AdaptiveCards/pull/632#discussion_r140325996) [](ancestors = 140325996)

Reply:
What I mean to say is that I don't think we need the Internal classes at all. The getCard and getHostConfig methods can be on the ACOAdaptiveCards and ACOHostConfig interfaces themselves.

They can be put into private categories to avoid exposing them publicly.

---
In reply to: [140346363](https://github.com/Microsoft/AdaptiveCards/pull/632#discussion_r140346363) [](ancestors = 140346363,140325996)

Reply:
that's great suggestion; thanks

---
In reply to: [140379817](https://github.com/Microsoft/AdaptiveCards/pull/632#discussion_r140379817) [](ancestors = 140379817,140346363,140325996)